src/ast/passes/semantic_analyser.h

```diff
@@ -110,8 +110,9 @@ class SemanticAnalyser : public Visitor {
   std::ostream &out_;
   std::ostringstream err_;
   int pass_;
-  const int num_passes_ = 10;
   bool listing_;
+  bool is_final_pass_ = false;
+  int num_unresolved_;
```

Comment:
Building on above comment, break these two vars out into its own grouping and comment to not directly modify these?

Reply:
Would be great to find some way to have an "even private-er" way to hide these vars. A small helper class?

Reply:
Also defensive `num_unresolved_ = 0` would be good